src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs

```diff
@@ -641,15 +645,23 @@ private void AssignCancellationToken(CancellationToken cancellationToken, Task a
                         if (antecedent == null)
                         {
                             // if no antecedent was specified, use this task's reference as the cancellation state object
-                            ctr = cancellationToken.UnsafeRegister(s_taskCancelCallback, this);
+                            ctr = cancellationToken.UnsafeRegister(t => ((Task)t).InternalCancel(false), this);
                         }
                         else
                         {
                             // If an antecedent was specified, pack this task, its antecedent and the TaskContinuation together as a tuple 
                             // and use it as the cancellation state object. This will be unpacked in the cancellation callback so that 
                             // antecedent.RemoveCancellation(continuation) can be invoked.
-                            ctr = cancellationToken.UnsafeRegister(s_taskCancelCallback,
-                                                                              new Tuple<Task, Task, TaskContinuation>(this, antecedent, continuation));
+                            ctr = cancellationToken.UnsafeRegister(t => {
+                                var tuple = (Tuple<Task, Task, TaskContinuation>) t;
+            
+                                Task targetTask = tuple.Item1;
+                                Task antecedentTask = tuple.Item2;
+
```

Comment:
I'm curious why you removed the local for tuple.Item3 but not for Item1 or Item2?

Reply:
There was name shadowing issue with `continuation` method parameter and I didn't think a new name like `continuation2` would add much